src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -1707,32 +1718,59 @@ internal void PerformUserCancellation()
             if (!ReadingPrependedMessagesMRE.Wait(0))
                 return;
 
-            _userCancellationRequested = true;
-
-            if (AttemptPostgresCancellation && SupportsPostgresCancellation)
-            {
-                var cancellationTimeout = Settings.CancellationTimeout;
-                if (PerformPostgresCancellation() && cancellationTimeout >= 0)
-                {
-                    if (cancellationTimeout > 0)
-                    {
-                        ReadBuffer.Timeout = TimeSpan.FromMilliseconds(cancellationTimeout);
-                        ReadBuffer.Cts.CancelAfter(cancellationTimeout);
-                    }
+            PerformUserCancellation();
+        }
+        finally
+        {
+            Monitor.Exit(CancelLock);
```

Comment:
Shouldn't be necessary here? CancelLock isn't taken here afaik?

Reply:
It should as we do take it just after the first `if`

Reply:
Ah I missed it https://github.com/npgsql/npgsql/pull/5774/files#diff-b45ed9d534512c2b95cd6aafbdceff63d87439bf00b57a605d637cc64785e438L1704